pkg/webhook/admission.go

```diff
@@ -0,0 +1,262 @@
+/*
+Copyright 2019 The OpenEBS Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhook
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	clientset "github.com/openebs/maya/pkg/client/generated/clientset/internalclientset"
+
+	"github.com/golang/glog"
+	"k8s.io/api/admission/v1beta1"
+	admissionregistrationv1beta1 "k8s.io/api/admissionregistration/v1beta1"
+	v1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+)
+
+var (
+	runtimeScheme = runtime.NewScheme()
+	codecs        = serializer.NewCodecFactory(runtimeScheme)
+	deserializer  = codecs.UniversalDeserializer()
+
+	// (https://github.com/kubernetes/kubernetes/issues/57982)
+	defaulter = runtime.ObjectDefaulter(runtimeScheme)
+)
+
+var (
+	ignoredNamespaces = []string{
+		metav1.NamespaceSystem,
+		metav1.NamespacePublic,
+	}
+	requiredLabels = []string{
+		nameLabel,
+		instanceLabel,
+		versionLabel,
+		componentLabel,
+		partOfLabel,
+		managedByLabel,
+	}
+	addLabels = map[string]string{
+		nameLabel:      NA,
+		instanceLabel:  NA,
+		versionLabel:   NA,
+		componentLabel: NA,
+		partOfLabel:    NA,
+		managedByLabel: NA,
+	}
+)
+
+const (
+	admissionWebhookAnnotationValidateKey = "admission-webhook-openebs.io/validate"
+	admissionWebhookAnnotationMutateKey   = "admission-webhook-openebs.io/mutate"
+	admissionWebhookAnnotationStatusKey   = "admission-webhook-openebs.io/status"
+
+	nameLabel      = "app.kubernetes.io/name"
+	instanceLabel  = "app.kubernetes.io/instance"
+	versionLabel   = "app.kubernetes.io/version"
+	componentLabel = "app.kubernetes.io/component"
+	partOfLabel    = "app.kubernetes.io/part-of"
+	managedByLabel = "app.kubernetes.io/managed-by"
+
+	NA = "not_available"
+)
+
+type WebhookServer struct {
+	Server *http.Server
+
+	// clientset is a openebs custom resource package generated for custom API group.
+	Clientset clientset.Interface
+}
+
+// Webhook Server parameters
+type WhSvrParameters struct {
+	Port     int    // webhook server port
+	CertFile string // path to the x509 certificate for https
+	KeyFile  string // path to the x509 private key matching `CertFile`
+}
+
+type patchOperation struct {
```

Comment:
I believe @sonasingh46 had earlier put some effort in separating this struct into some other package. I see this struct is used repeatedly and is redundant. This is the right time to extract it out and make use of common struct.